sys/usb/usbus/usbus_fmt.c

```diff
@@ -0,0 +1,279 @@
+        usbus_control_slicer_put_bytes(usbus, (uint8_t *)&usb_ep,
+                                       sizeof(usb_descriptor_endpoint_t));
+        _hdrs_fmt_additional(usbus, ep->hdr_gen);
+        len += usb_ep.length;
+        /* iterate to next endpoint */
+        ep = ep->next;
+    }
+    return len;
+}
+
+static void _hdrs_fmt_iface(usbus_interface_t *iface,
+                            usb_descriptor_interface_t *usb_iface)
+{
+    memset(usb_iface, 0, sizeof(usb_descriptor_interface_t));
+    usb_iface->length = sizeof(usb_descriptor_interface_t);
+    usb_iface->type = USB_TYPE_DESCRIPTOR_INTERFACE;
+    usb_iface->interface_num = iface->idx;
+    usb_iface->class = iface->class;
+    usb_iface->subclass = iface->subclass;
+    usb_iface->protocol = iface->protocol;
+}
+
+static size_t _hdrs_fmt_iface_alts(usbus_t *usbus, usbus_interface_t *iface)
+{
+    size_t len = 0;
+    uint8_t alts = 1;
+
+    for (usbus_interface_alt_t *alt = iface->alts;
+         alt;
+         alt = alt->next) {
+        usb_descriptor_interface_t usb_iface;
+        _hdrs_fmt_iface(iface, &usb_iface);
+        usb_iface.alternate_setting = alts++;
+        usb_iface.num_endpoints = _num_endpoints_alt(alt);
+        usbus_control_slicer_put_bytes(usbus, (uint8_t *)&usb_iface,
+                                       sizeof(usb_descriptor_interface_t));
+        len += _hdrs_fmt_additional(usbus, alt->hdr_gen);
+        len += _hdrs_fmt_endpoints(usbus, alt->ep);
+    }
+    return len;
+}
+
+static size_t _hdrs_fmt_ifaces(usbus_t *usbus)
+{
+    size_t len = 0;
+
+    for (usbus_interface_t *iface = usbus->iface;
+         iface;
+         iface = iface->next) {
+        usb_descriptor_interface_t usb_iface;
+        _hdrs_fmt_iface(iface, &usb_iface);
+        usb_iface.num_endpoints = _num_endpoints(iface);
+        if (iface->descr) {
+            usb_iface.idx = iface->descr->idx;
+        }
+        else {
+            usb_iface.idx = 0;
+        }
+        usbus_control_slicer_put_bytes(usbus, (uint8_t *)&usb_iface,
+                                       sizeof(usb_descriptor_interface_t));
+        len += sizeof(usb_descriptor_interface_t);
+        len += _hdrs_fmt_additional(usbus, iface->hdr_gen);
+        len += _hdrs_fmt_endpoints(usbus, iface->ep);
+        len += _hdrs_fmt_iface_alts(usbus, iface);
+    }
+    return len;
+}
+
+size_t usbus_fmt_hdr_conf(usbus_t *usbus)
+{
+    size_t len = 0;
+    usb_descriptor_configuration_t conf;
+
+    memset(&conf, 0, sizeof(usb_descriptor_configuration_t));
+    conf.length = sizeof(usb_descriptor_configuration_t);
+    conf.type = USB_TYPE_DESCRIPTOR_CONFIGURATION;
+    conf.total_length = sizeof(usb_descriptor_configuration_t);
+    conf.val = 1;
+    conf.attributes = USB_CONF_ATTR_RESERVED;
+    if (USB_CONFIG_SELF_POWERED) {
+        conf.attributes |= USB_CONF_ATTR_SELF_POWERED;
+    }
+    /* Todo: upper bound */
+    conf.max_power = USB_CONFIG_MAX_POWER / 2;
```

Comment:
why / 2 ?

Reply:
the bMaxPower (table 9-10 from the usb 2.0 spec) is in 2mA increments. I prefer to have the define in actual mA for convenience.

Reply:
I can put a comment here to explain

Reply:
I would be very happy !